File: backend/app/core/article_keys.py
from __future__ import annotations

import re
import unicodedata


def normalize_article_title(title: str) -> str:
    normalized = unicodedata.normalize('NFKC', title or '').strip().lower()
    normalized = re.sub(r'[^\w\s-]', '', normalized, flags=re.UNICODE)
    normalized = re.sub(r'[_\s-]+', '-', normalized, flags=re.UNICODE).strip('-')
    return normalized or 'article'


def build_article_key(issue_slug: str, article_title: str, article_index: int) -> str:
    safe_index = max(0, int(article_index))
    return f'{issue_slug}::{normalize_article_title(article_title)}::{safe_index}'

File: frontend/lib/article-keys.ts
export function normalizeArticleTitle(title: string) {
  const normalized = title.normalize('NFKC').trim().toLowerCase();
  const stripped = normalized
    .replace(/[^\p{L}\p{N}\s_-]+/gu, '')
    .replace(/[_\s-]+/g, '-')
    .replace(/^-+|-+$/g, '');
  return stripped || 'article';
}

export function buildArticleKey(issueSlug: string, articleTitle: string, articleIndex: number) {
  return `${issueSlug}::${normalizeArticleTitle(articleTitle)}::${Math.max(0, articleIndex)}`;
}

export function buildArticleDomId(articleKey: string) {
  return `article-${articleKey.replace(/[^a-zA-Z0-9_-]+/g, '-')}`;
}
